File: packages/i18nify-js/src/modules/geo/formatAddressWithFormat.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import formatAddress from './formatAddress';
import { getAddressInfo } from './getAddressInfo';
import { AddressCodeType, AddressType, AddressComponents } from './types';

/**
 * Formats address components using the country-specific template from country metadata.
 * Empty component fields substitute as blank strings; lines that become empty
 * after substitution are removed from the output.
 */
const formatAddressWithFormat = (
  countryCode: string,
  components: AddressComponents,
): string => {
  const trimmed = countryCode.trim();
  if (trimmed === '') {
    throw new Error('formatAddressWithFormat: country code must not be empty.');
  }
  const code = trimmed.toUpperCase() as AddressCodeType;

  const addressInfo = getAddressInfo(code) as AddressType | null;

  if (!addressInfo) {
    throw new Error(
      `formatAddressWithFormat: address format for country code "${code}" not found.`,
    );
  }

  // Reuse the base formatter so placeholder substitution and blank-line cleanup
  // stay defined in exactly one place for the JS geo module.
  return formatAddress(addressInfo.address_template, components);
};

export default withErrorBoundary<typeof formatAddressWithFormat>(
  formatAddressWithFormat,
);

File: packages/i18nify-go/modules/geo/format_address_with_format.go
package geo

import (
	"fmt"
	"strings"
)

// AddressComponents holds address field values for template substitution.
// All fields are optional; empty fields substitute as blank strings and
// lines that become blank after substitution are dropped from the output.
type AddressComponents struct {
	Name          string
	Organization  string
	StreetAddress string
	City          string
	State         string
	Zip           string
	District      string
	SortingCode   string
}

// FormatAddressWithFormat formats address components using the country-specific template for countryCode.
// Templates are loaded from embedded country metadata at startup (see geo.go init).
// Each {placeholder} in the template is replaced with the matching field; blank lines in the result are removed.
func FormatAddressWithFormat(countryCode string, components AddressComponents) (string, error) {
	if strings.TrimSpace(countryCode) == "" {
		return "", fmt.Errorf("formatAddressWithFormat: country code must not be empty")
	}

	code := strings.ToUpper(strings.TrimSpace(countryCode))

	countryMeta, exists := cachedCountryMetadata.GetMetadataInformation()[code]
	if !exists || countryMeta == nil || countryMeta.GetAddressFormat() == "" {
		return "", fmt.Errorf("formatAddressWithFormat: address format for country code %q not found", code)
	}

	// Reuse the base formatter so template substitution and blank-line cleanup
	// remain defined in one place for the Go geo module.
	return FormatAddress(countryMeta.GetAddressFormat(), components)
}
